File: vignettes/24_compaction/python/24_compaction.py
"""
Message Compaction — Managing context windows (Python)

This sample demonstrates the compaction strategies available in the Python
agent-framework for reducing message history when conversations exceed the
model's context window.  It mirrors the Julia vignette 24_compaction.

Prerequisites:
  - pip install agent-framework
"""

import asyncio

from agent_framework import Message, Content
from agent_framework._compaction import (
    SlidingWindowStrategy,
    SelectiveToolCallCompactionStrategy,
    ToolResultCompactionStrategy,
    SummarizationStrategy,
    TokenBudgetComposedStrategy,
    TruncationStrategy,
    annotate_message_groups,
    extend_compaction_messages,
    included_messages,
    included_token_count,
)


# ── Helper: build a long conversation ────────────────────────────────────

def build_conversation(n_turns: int = 50) -> list[Message]:
    """Create a synthetic conversation with `n_turns` user/assistant pairs."""
    messages: list[Message] = [
        Message("system", ["You are a helpful assistant."]),
    ]
    for i in range(1, n_turns + 1):
        messages.append(
            Message("user", [f"Question {i}: {'x' * 200}"])
        )
        messages.append(
            Message("assistant", [f"Answer {i}: {'y' * 300}"])
        )
    return messages


def build_tool_conversation() -> list[Message]:
    """Create a conversation that includes tool calls and results."""
    return [
        Message("system", ["You are a calculator agent."]),
        Message("user", ["What is 2+2?"]),
        Message(
            "assistant",
            [
                Content.from_function_call(
                    call_id="call_1",
                    name="add",
                    arguments='{"a": 2, "b": 2}',
                )
            ],
        ),
        Message(
            "tool",
            [
                Content.from_function_result(
                    call_id="call_1",
                    result="4",
                )
            ],
        ),
        Message("assistant", ["2+2 = 4"]),
        Message("user", ["And 10*3?"]),
        Message(
            "assistant",
            [
                Content.from_function_call(
                    call_id="call_2",
                    name="multiply",
                    arguments='{"a": 10, "b": 3}',
                )
            ],
        ),
        Message(
            "tool",
            [
                Content.from_function_result(
                    call_id="call_2",
                    result="30",
                )
            ],
        ),
        Message("assistant", ["10*3 = 30"]),
    ]


async def main() -> None:
    messages = build_conversation(n_turns=50)
    print(f"Conversation has {len(messages)} messages")

    # ── Sliding Window ───────────────────────────────────────────────────
    print("\n=== Sliding Window ===")
    sliding = SlidingWindowStrategy(keep_last_groups=6)
    annotate_message_groups(messages)
    await sliding(messages)
    kept = included_messages(messages)
    print(f"Kept {len(kept)} of {len(messages)} messages")

    # ── Truncation ───────────────────────────────────────────────────────
    print("\n=== Truncation ===")
    messages = build_conversation(n_turns=50)
    truncation = TruncationStrategy(max_n=20, compact_to=10)
    annotate_message_groups(messages)
    await truncation(messages)
    kept = included_messages(messages)
    print(f"Kept {len(kept)} of {len(messages)} messages")

    # ── Selective Tool Call ───────────────────────────────────────────────
    print("\n=== Selective Tool Call ===")
    tool_messages = build_tool_conversation()
    selective = SelectiveToolCallCompactionStrategy(keep_last_tool_call_groups=1)
    annotate_message_groups(tool_messages)
    await selective(tool_messages)
    kept = included_messages(tool_messages)
    print(f"Kept {len(kept)} of {len(tool_messages)} messages")

    # ── Tool Result Replacement ──────────────────────────────────────────
    print("\n=== Tool Result Replacement ===")
    tool_messages = build_tool_conversation()
    tool_result = ToolResultCompactionStrategy(keep_last_tool_call_groups=1)
    annotate_message_groups(tool_messages)
    await tool_result(tool_messages)
    kept = included_messages(tool_messages)
    print(f"Kept {len(kept)} of {len(tool_messages)} messages")
    for msg in kept:
        if msg.role == "tool":
            text = msg.text if hasattr(msg, 'text') else str(msg.contents)
            print(f"  Tool result: {text}")

    # ── Pipeline (composed strategies) ───────────────────────────────────
    print("\n=== Pipeline (TokenBudgetComposedStrategy) ===")
    messages = build_conversation(n_turns=50)

    # A simple char-based tokenizer for demonstration
    class CharTokenizer:
        def count_tokens(self, text: str) -> int:
            return max(1, len(text) // 4)

    tokenizer = CharTokenizer()
    pipeline = TokenBudgetComposedStrategy(
        token_budget=2048,
        tokenizer=tokenizer,
        strategies=[
            SelectiveToolCallCompactionStrategy(keep_last_tool_call_groups=2),
            SlidingWindowStrategy(keep_last_groups=6),
        ],
    )
    annotate_message_groups(messages, tokenizer=tokenizer)
    await pipeline(messages)
    kept = included_messages(messages)
    print(f"Pipeline kept {len(kept)} of {len(messages)} messages")

    # ── Summary ──────────────────────────────────────────────────────────
    print("\n=== Strategy Summary ===")
    print("SlidingWindowStrategy     — Keep N most recent groups")
    print("TruncationStrategy        — Drop oldest until under budget")
    print("SelectiveToolCallStrategy — Remove old tool call/result pairs")
    print("ToolResultCompaction      — Replace old tool results with summary")
    print("SummarizationStrategy     — Summarize old messages via LLM")
    print("TokenBudgetComposed       — Chain strategies in a pipeline")


if __name__ == "__main__":
    asyncio.run(main())

File: vignettes/24_compaction/csharp/24_compaction.cs
// Message Compaction — Managing context windows (C#)
//
// This sample demonstrates the compaction strategies available in the C#
// agent-framework for reducing message history when conversations exceed
// the model's context window.  It mirrors the Julia vignette 24_compaction.
//
// Prerequisites:
//   - dotnet restore

using Microsoft.Agents.AI;
using Microsoft.Agents.AI.Compaction;
using Microsoft.Extensions.AI;
using Microsoft.Extensions.Logging;
using Microsoft.Extensions.Logging.Abstractions;

var logger = NullLogger.Instance;

// ── Build a long conversation ────────────────────────────────────────────

List<ChatMessage> BuildConversation(int turns = 50)
{
    var messages = new List<ChatMessage>
    {
        new(ChatRole.System, "You are a helpful assistant.")
    };
    for (int i = 1; i <= turns; i++)
    {
        messages.Add(new ChatMessage(ChatRole.User, $"Question {i}: {new string('x', 200)}"));
        messages.Add(new ChatMessage(ChatRole.Assistant, $"Answer {i}: {new string('y', 300)}"));
    }
    return messages;
}

List<ChatMessage> BuildToolConversation()
{
    return
    [
        new(ChatRole.System, "You are a calculator agent."),
        new(ChatRole.User, "What is 2+2?"),
        new(ChatRole.Assistant,
        [
            new FunctionCallContent("call_1", "add", new Dictionary<string, object?> { ["a"] = 2, ["b"] = 2 })
        ]),
        new(ChatRole.Tool, [new FunctionResultContent("call_1", "4")]),
        new(ChatRole.Assistant, "2+2 = 4"),
        new(ChatRole.User, "And 10*3?"),
        new(ChatRole.Assistant,
        [
            new FunctionCallContent("call_2", "multiply", new Dictionary<string, object?> { ["a"] = 10, ["b"] = 3 })
        ]),
        new(ChatRole.Tool, [new FunctionResultContent("call_2", "30")]),
        new(ChatRole.Assistant, "10*3 = 30"),
    ];
}

// ── Sliding Window ───────────────────────────────────────────────────────

Console.WriteLine("=== Sliding Window ===");
{
    var messages = BuildConversation(50);
    var strategy = new SlidingWindowCompactionStrategy(
        trigger: CompactionTriggers.TurnsExceed(10),
        target: CompactionTriggers.TurnsExceed(6));

    var compacted = await CompactionProvider.CompactAsync(
        strategy, messages, logger);

    Console.WriteLine($"Kept {compacted.Count()} of {messages.Count} messages");
}

// ── Truncation ───────────────────────────────────────────────────────────

Console.WriteLine("\n=== Truncation ===");
{
    var messages = BuildConversation(50);
    var strategy = new TruncationCompactionStrategy(
        trigger: CompactionTriggers.MessagesExceed(20),
        target: CompactionTriggers.MessagesExceed(10));

    var compacted = await CompactionProvider.CompactAsync(
        strategy, messages, logger);

    Console.WriteLine($"Kept {compacted.Count()} of {messages.Count} messages");
}

// ── Pipeline (chained strategies) ────────────────────────────────────────

Console.WriteLine("\n=== Pipeline ===");
{
    var messages = BuildConversation(50);
    // PipelineCompactionStrategy takes an ordered sequence of strategies.
    // Each child strategy evaluates its own trigger independently.
    var pipeline = new PipelineCompactionStrategy(
        new CompactionStrategy[]
        {
            new TruncationCompactionStrategy(
                trigger: CompactionTriggers.MessagesExceed(40),
                minimumPreservedGroups: 20),
            new SlidingWindowCompactionStrategy(
                trigger: CompactionTriggers.TurnsExceed(10),
                minimumPreservedTurns: 4),
        });

    var compacted = await CompactionProvider.CompactAsync(
        pipeline, messages, logger);

    Console.WriteLine($"Pipeline kept {compacted.Count()} of {messages.Count} messages");
}

// ── Tool-Call Compaction (atomic group handling) ─────────────────────────

Console.WriteLine("\n=== Tool-Call Compaction ===");
{
    // Tool-call/result pairs are grouped atomically — they are always
    // removed together, never split across a compaction boundary.
    var messages = BuildToolConversation();
    var strategy = new TruncationCompactionStrategy(
        trigger: CompactionTriggers.MessagesExceed(4),
        minimumPreservedGroups: 2);

    var compacted = await CompactionProvider.CompactAsync(
        strategy, messages, logger);

    Console.WriteLine($"Kept {compacted.Count()} of {messages.Count} messages (tool pairs stay atomic)");
}

// ── Strategy Summary ─────────────────────────────────────────────────────

Console.WriteLine("\n=== Strategy Summary ===");
Console.WriteLine("TruncationCompactionStrategy   — Remove oldest non-system groups");
Console.WriteLine("SlidingWindowCompactionStrategy — Keep N most recent turns");
Console.WriteLine("PipelineCompactionStrategy     — Chain strategies sequentially");
Console.WriteLine("CompactionTriggers             — Predicate factory (tokens, messages, turns)");
Console.WriteLine("CompactionMessageIndex         — Low-level group/message index");
